### PR TITLE
DuckDB as memory storage

### PR DESCRIPTION
**Pull request by colombod**
### Motivation and Context

DuckDB is becoming a popular option for in memory OLAP storage. This pr contributes a MemoryStorage implementation on DuckDB using both tile and in memory setup.

### Description

DuckDB memory storage implementation

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    public static async Task<DuckDBMemoryStore> ConnectAsync(string filename,
+        CancellationToken cancellationToken = default)
+    {
+        var memoryStore = new DuckDBMemoryStore($"Data Source={filename}");
+        await memoryStore._dbConnection.OpenAsync(cancellationToken).ConfigureAwait(false);
+        await memoryStore._dbConnector.CreateTableAsync(memoryStore._dbConnection, cancellationToken).ConfigureAwait(false);
+        return memoryStore;
+    }
+
+    public static async Task<DuckDBMemoryStore> ConnectAsync(
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs:41]: Add xml docs here too.

File: dotnet/src/Connectors/Connectors.UnitTests/Memory/DuckDB/DuckDBMemoryStoreTests.cs [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+using Microsoft.SemanticKernel.AI.Embeddings;
+using Microsoft.SemanticKernel.Connectors.Memory.DuckDB;
+using Microsoft.SemanticKernel.Connectors.Memory.Sqlite;
+using Microsoft.SemanticKernel.Memory;
+using Xunit;
+
+namespace SemanticKernel.Connectors.UnitTests.Memory.DuckDB;
+
+/// <summary>
+/// Unit tests of <see cref="SqliteMemoryStore"/>.
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.UnitTests/Memory/DuckDB/DuckDBMemoryStoreTests.cs:17]:
```suggestion
/// Unit tests of <see cref="DuckDBMemoryStore"/>.
```

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/Connectors.Memory.DuckDB.csproj [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    <PackageReference Include="DuckDB.NET.Data.Full" />
+    <PackageReference Include="DuckDB.NET.Data" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\SemanticKernel\SemanticKernel.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <InternalsVisibleTo Include="SemanticKernelConnectorsSqliteTests"/>
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.DuckDB/Connectors.Memory.DuckDB.csproj:29]: Update this.

File: dotnet/src/Connectors/Connectors.UnitTests/Memory/DuckDB/DuckDBMemoryStoreTests.cs [hunk 4]
```diff
@@ -0,0 +1,658 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.AI.Embeddings;
+using Microsoft.SemanticKernel.Connectors.Memory.DuckDB;
+using Microsoft.SemanticKernel.Connectors.Memory.Sqlite;
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.UnitTests/Memory/DuckDB/DuckDBMemoryStoreTests.cs:10]: Shouldn't be needed.

**Review by lemillermicrosoft — CHANGES_REQUESTED** [review]
Looks great overall! Haven't tested yet. Left a few comments for now.

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/Connectors.Memory.DuckDB.csproj (hunk 3 above)
Comment (colombod) [reply dotnet/src/Connectors/Connectors.Memory.DuckDB/Connectors.Memory.DuckDB.csproj:29]: removed

File: dotnet/src/Connectors/Connectors.UnitTests/Memory/DuckDB/DuckDBMemoryStoreTests.cs (hunk 4 above)
Comment (colombod) [reply dotnet/src/Connectors/Connectors.UnitTests/Memory/DuckDB/DuckDBMemoryStoreTests.cs:10]: removed sql lite namespace import

File: dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs (hunk 1 above)
Comment (colombod) [reply dotnet/src/Connectors/Connectors.Memory.DuckDB/DuckDBMemoryStore.cs:41]: done

**Comment by colombod** [comment]
Addressed comments

**Comment by colombod** [comment]
> Looks great overall! Haven't tested yet. Left a few comments for now.

I think I have addressed the comments

**Review by lemillermicrosoft — APPROVED** [review]

**Review by awharrison-28 — APPROVED** [review]

**Pull request merged**